### ISSUE TITLE
Matrix array param is not retrievable by ParamMap#getAll function.

### ISSUE THREAD

**Issue by zoltan-gal** [angular/angular#19179]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

If an URL is built with matrix parameters which contain array value, it is not retrievable as an elements of  array by the ParamMap.getAll function. It provides an array with only one single value in which the original values are concatenated instead of multiple elements.

For example:
- Given a navigation like this: this.router.navigate(['example', { foo: ['bar', 'baz'] } ]);
- The URL will be something like this: /example;foo=bar,baz
- Then I want to get back the foo values through the ActivatedRoute: this.activatedRoute.snapshot.paramMap.getAll('foo')
- I get this: ['bar,baz'] instead of ['bar', 'baz']

## Expected behavior

- getAll function should provide an elements of array (['foo', 'bar']) instead of an array with single element (['foo,bar']).

## Environment

<pre><code>
Angular version: 4.3.1

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform: Windows 

Others:

</code></pre>

**Comment by mariusdavidschneider** [comment]
Ran into the same thing. I opted for QueryParams instead, as these seem to pass arrays just fine - but that opens an entire other can of worms...

**Comment by bennylava10100** [comment]
Do we have an ETA when this will be fixed? Still an issue in Angular 5

**Comment by tbunique** [comment]
@bennylava10100 Do you have an example of it working with router.navigate or routerLink?  
I tried both and I get the same result as in the first post.

**Comment by bennylava10100** [comment]
@tbunique sorry for the late reply, no I don't have an example of it working, it seems the issue persists in ng7 as in the first post.

**Comment by ayyash** [comment]
Angular 7.2 here, still getting a string, I'm not sure if this is the intended behavior though

**Comment by DibyodyutiMondal** [comment]
I am new to Angular and Typescript and ran into the same problem. After a lot of reading and veritably frustrated with not finding a default for a feature that I now admit I took for granted with frameworks like asp.net, and misled due to expectations that canLoad() will work like canActivate(), I am publishing my solution on this thread with the hope that it saves others valuable time

Relevant excerpts from AuthGuard.ts:

```
@Injectable({
  providedIn: 'root'
})
export class AuthGuard implements CanLoad {

  constructor(
    //inject router for navigation
    private router: Router,
    //other required injections for your case
  ) { }

  canLoad(route: Route, segments: UrlSegment[]): boolean | Observable<boolean> | Promise<boolean> {
    //----implement your authorization logic----//
   
   //when you need to navigate away to login:
  this.router.navigate(['/Login'], {
        queryParams: {
          //I'm using 'returnUrl' here because in my login component, I am using
          //a query parameter with that exact name to redirect after successful authentication
          returnUrl: this.UrlSegmentsToUrl(segments)
        }
      });
}
```

```
private UrlSegmentsToUrl(segments: UrlSegment[]): string {
    //containing string for the result url
    let url: string = '';

    //for each urlSegment, add the path with a leading slash
    segments.forEach(function (value) {
      url = url + '/' + value.path;

      //for each matrix parameter corresponding to the current segment,
      //add the name and value(s) with a leading semicolon
      value.parameterMap.keys.forEach(function (mpk) {

        //get all the value of the current parameter. As of implementation,
        //getAll() returns an array with only one string having comma-separated values,
        //so no further processing needed, but if that changes in the future,
        // we'll need to make a single 'params' string using necessary logic
        let params = value.parameterMap.getAll(mpk)[0];

        url = url + ';' + mpk + '=' + params
      });
    });

    //add any query parameters if present
    let qpindex = window.location.href.indexOf("?");
    if (qpindex != -1) {
      url = url + window.location.href.substr(qpindex);
    }

    return url;
  }
```

I read somewhere that router state is not accessible at the stage when canLoad() is called for routes, so the only choice left was to parse the segments into a string ourselves. It works for both matrix parameters and query parameters, and even if both are used simultaneously. 

As you can see, this method depends on the browser through window.location.href and I am unsure if that is a good thing or a bad thing. However, I tried it with anchor-tag links using both 'deferred' and 'eager' url update strategy, as well as directly typing urls into the browser and they worked
```
//For those who don't know what url update strategy is and how to configure it
@NgModule({
  declarations: [ ],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    RouterModule,
    RouterModule.forRoot(
      routes,
      {
        preloadingStrategy: PreloadAllModules,
        anchorScrolling: "enabled",
        scrollPositionRestoration: "enabled",
        urlUpdateStrategy: 'eager'
      }),
    HttpClientModule,
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

However, it is unknown to me what effects configuring the 'paramsInheritanceStrategy' will have on the results of this function. It would be appreciated if those more familiar with this issue clarify that for us.

The UrlSegmentsToUrl() method is heavily expanded for clarity and probably unoptimized because I am a new at Typescript and Angular. Improvements will be appreciated.

**Comment by benneq** [comment]
Angular 8 here, still the same issue.

I'm not sure if matrix params are supposed to work with arrays at all. Here are my observations:

- If you have something like `;foo=1,2,3` in your URL, it's not clear if it's a single string `'1,2,3'` or if it's an array of values `['1', '2', '3']`
- Though it should be better to use `;foo=1;foo=2;foo=3`, but angular replaces this automatically with `;foo=3`. (why?!)

**Comment by MaKCbIMKo** [comment]
Faced this case as well (with Angular 8.2).

Is there any progress on it?

Documentation says it should work properly with matrix/optional params. And I agree with @benneq observations.

So far if I want to use matrix/optional params I need to stick with comma-separated values and parse them manually into array. But would be nice to have this case handled properly out of the box (as documentation says it should be).

**Comment by ericsnap** [comment]
FWIW, I am seeing _inconsistent_ behavior. Still trying to figure out how to get one vs. the other, but I sometimes get `['foo','bar']` and sometimes `['foo,bar']`.

**Comment by MarkoSulamagi** [comment]
The issue is still relevant (A9)

**Comment by scote** [comment]
I have same issue in A10.

**Comment by enua** [comment]
A12, same issue :(

**Comment by drtomato** [comment]
A13, still the same issue.

**Comment by zbogdanov** [comment]
This is a bug for 5 years now. @IgorMinar I kindly ask to prioritize this, it does not look good from a community feedback perspective

**Comment by stas-solc** [comment]
I can confirm version 18.2.1 still has this problem. Without support of matrix array params, _paramsMap_ observable has little meaning...

```
<a [routerLink]="['/matrix-params', { ids: [1,2,3] }]">Matrix params</a>
this.route.paramMap.subscribe(paramMap => {
   console.log(paramMap.getAll('ids'));
})
```
/matrix-params;ids=1,2,3 - OK
result is single string: ['1,2,3'] - FAIL  expected array [1,2,3]

**Comment by atscott** [comment]
I don't think we can move forward with this change as altering how matrix parameters are serialized or parsed is not worth the breaking change to URL generation and matching.

You can handle array values in userland by joining when writing and splitting when reading:

### Writing
```ts
this.router.navigate(['/example', { foo: ['bar', 'baz'].join(',') }]);
// Produces: /example;foo=bar,baz
```

### Reading
```ts
// Via Snapshot:
const foo = this.route.snapshot.paramMap.get('foo')?.split(',') ?? [];

// Or with Observables / toSignal:
readonly foo$ = this.route.paramMap.pipe(
  map(params => params.get('foo')?.split(',') ?? [])
);
```

Closing as this can be handled directly in user code.
